File: elixir-elasticemail-examples/phoenix_app/lib/phoenix_app_web/controllers/double_optin_controller.ex
defmodule PhoenixAppWeb.DoubleOptinController do
  use Phoenix.Controller, formats: [:json]
  import PhoenixAppWeb.Helpers
  alias ElasticEmail.Env

  def subscribe(conn, %{"email" => email} = params) when is_binary(email) and email != "" do
    name = to_string(params["name"] || "")
    first_name = name |> String.split(" ") |> List.first()

    confirm_url =
      "#{Env.public_url()}/double-optin/confirm?email=#{URI.encode_www_form(email)}&token=#{hmac(email)}"

    greeting = if name == "", do: "Welcome!", else: "Welcome, #{name}!"

    html = """
    <div style="text-align: center; padding: 40px 20px; font-family: Arial, sans-serif;">
      <h1>#{greeting}</h1>
      <p>Please confirm your subscription to our newsletter.</p>
      <a href="#{confirm_url}" style="background-color: #18181b; color: #fff; padding: 12px 32px; border-radius: 6px; text-decoration: none; font-weight: bold; display: inline-block;">Confirm Subscription</a>
    </div>
    """

    # Stored as Transactional so it receives the confirmation but no campaigns yet
    with {:ok, _} <-
           ElasticEmail.post("/contacts", [
             %{"Email" => email, "FirstName" => first_name, "Status" => "Transactional"}
           ]),
         {:ok, data} <-
           ElasticEmail.send_transactional(%{
             "Recipients" => %{"To" => [email]},
             "Content" => %{
               "From" => Env.from(),
               "Subject" => "Confirm your subscription",
               "Body" => [%{"ContentType" => "HTML", "Content" => html}]
             }
           }) do
      json(conn, %{success: true, message: "Confirmation email sent", messageId: data["MessageID"]})
    else
      error -> api_error(conn, error)
    end
  end

  def subscribe(conn, _params) do
    conn
    |> put_status(400)
    |> json(%{error: "Missing required field: email"})
  end

  def confirm(conn, params) do
    email = to_string(params["email"] || "")
    token = to_string(params["token"] || "")

    if email == "" or not Plug.Crypto.secure_compare(hmac(email), token) do
      conn
      |> put_status(400)
      |> json(%{error: "Invalid confirmation link"})
    else
      list_name = Env.list_name()

      case add_to_list(list_name, email) do
        {:ok, _} ->
          case Env.confirm_redirect_url() do
            nil -> json(conn, %{confirmed: true, email: email, list: list_name})
            url -> redirect(conn, external: url)
          end

        error ->
          api_error(conn, error)
      end
    end
  end

  # Click-tracking based confirmation: create a webhook for Clicked events pointing here.
  def webhook(conn, params) do
    if token_ok?(params["token"]) do
      confirm_from_click(conn, params)
    else
      unauthorized(conn)
    end
  end

  defp confirm_from_click(conn, event) do
    status = sanitize(event["status"])
    target = sanitize(event["target"])
    recipient = sanitize(event["to"])

    if status != "Clicked" or not String.contains?(target, "/double-optin/confirm") do
      json(conn, %{received: true, status: status, message: "Event ignored"})
    else
      case add_to_list(Env.list_name(), recipient) do
        {:ok, _} -> json(conn, %{received: true, confirmed: true, email: recipient})
        error -> api_error(conn, error)
      end
    end
  end

  defp add_to_list(list_name, email) do
    ElasticEmail.post("/lists/#{ElasticEmail.path_segment(list_name)}/contacts", %{
      "Emails" => [email]
    })
  end
end

File: nestjs-elasticemail-examples/typescript/src/double-optin/double-optin.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpCode,
  Inject,
  Post,
  Query,
  Res,
  UseGuards,
} from "@nestjs/common";
import type { Response } from "express";
import { ElasticEmailService } from "../elasticemail/elasticemail.service";
import { hmac, safeEqual, sanitize } from "../common/security";
import { TokenGuard } from "../common/token.guard";

type Fields = Record<string, string>;

@Controller("double-optin")
export class DoubleOptinController {
  constructor(@Inject(ElasticEmailService) private readonly ee: ElasticEmailService) {}

  @Post("subscribe")
  @HttpCode(200)
  async subscribe(@Body() body: { email?: string; name?: string }) {
    const { email, name = "" } = body ?? {};
    if (!email) {
      throw new BadRequestException("Missing required field: email");
    }

    const confirmUrl = `${this.ee.publicUrl}/double-optin/confirm?email=${encodeURIComponent(email)}&token=${hmac(email)}`;
    const greeting = name ? `Welcome, ${name}!` : "Welcome!";

    // Stored as Transactional so it receives the confirmation but no campaigns yet
    await this.ee.contacts.contactsPost([
      { Email: email, FirstName: name.split(" ")[0] || "", Status: "Transactional" },
    ]);

    const { data } = await this.ee.emails.emailsTransactionalPost({
      Recipients: { To: [email] },
      Content: {
        From: this.ee.from,
        Subject: "Confirm your subscription",
        Body: [
          {
            ContentType: "HTML",
            Content: `<div style="text-align: center; padding: 40px 20px; font-family: Arial, sans-serif;">
  <h1>${greeting}</h1>
  <p>Please confirm your subscription to our newsletter.</p>
  <a href="${confirmUrl}" style="background-color: #18181b; color: #fff; padding: 12px 32px; border-radius: 6px; text-decoration: none; font-weight: bold; display: inline-block;">Confirm Subscription</a>
</div>`,
          },
        ],
      },
    });

    return { success: true, message: "Confirmation email sent", messageId: data.MessageID };
  }

  @Get("confirm")
  async confirm(@Query("email") emailParam: unknown, @Query("token") token: unknown, @Res() res: Response) {
    const email = String(emailParam ?? "");
    if (!email || !safeEqual(token, hmac(email))) {
      throw new BadRequestException("Invalid confirmation link");
    }

    await this.ee.lists.listsByNameContactsPost(this.ee.listName, { Emails: [email] });
    if (this.ee.confirmRedirectUrl) {
      res.redirect(this.ee.confirmRedirectUrl);
      return;
    }
    res.json({ confirmed: true, email, list: this.ee.listName });
  }

  // Click-tracking based confirmation: create a webhook for Clicked events pointing here.
  // Elastic Email sends events as GET; a form POST is accepted too.
  @Get("webhook")
  @UseGuards(TokenGuard)
  webhookGet(@Query() query: Fields) {
    return this.handleClick({ ...query });
  }

  @Post("webhook")
  @HttpCode(200)
  @UseGuards(TokenGuard)
  webhookPost(@Query() query: Fields, @Body() body: Fields) {
    return this.handleClick({ ...query, ...(body ?? {}) });
  }

  private async handleClick(event: Fields) {
    if (event.status !== "Clicked" || !String(event.target ?? "").includes("/double-optin/confirm")) {
      return { received: true, status: sanitize(event.status), message: "Event ignored" };
    }

    await this.ee.lists.listsByNameContactsPost(this.ee.listName, { Emails: [event.to] });
    return { received: true, confirmed: true, email: sanitize(event.to) };
  }
}
